// @docusaurus/core/lib/index.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeTranslations = exports.writeHeadingIds = exports.swizzle = exports.start = exports.serve = exports.externalCommand = exports.deploy = exports.clear = exports.build = void 0;
const tslib_1 = require("tslib");
const build_1 = tslib_1.__importDefault(require("./commands/build"));
exports.build = build_1.default;
const clear_1 = tslib_1.__importDefault(require("./commands/clear"));
exports.clear = clear_1.default;
const deploy_1 = tslib_1.__importDefault(require("./commands/deploy"));
exports.deploy = deploy_1.default;
const external_1 = tslib_1.__importDefault(require("./commands/external"));
exports.externalCommand = external_1.default;
const serve_1 = tslib_1.__importDefault(require("./commands/serve"));
exports.serve = serve_1.default;
const start_1 = tslib_1.__importDefault(require("./commands/start"));
exports.start = start_1.default;
const swizzle_1 = tslib_1.__importDefault(require("./commands/swizzle"));
exports.swizzle = swizzle_1.default;
const writeHeadingIds_1 = tslib_1.__importDefault(require("./commands/writeHeadingIds"));
exports.writeHeadingIds = writeHeadingIds_1.default;
const writeTranslations_1 = tslib_1.__importDefault(require("./commands/writeTranslations"));
exports.writeTranslations = writeTranslations_1.default;
